src/main.cpp

```diff
@@ -5361,6 +5370,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
                     if (setMisbehaving.count(fromPeer))
                         continue;
+                    cs_orphancache.unlock(); // must maintain correct locking order
```

Comment:
Unlocking cs_orphancache here means that the orphanTx reference is carried across a unlock/lock of the guard.  So in theory some other thread could modify mapOrphanTransactions, invalidating the pointer under the orphanTx reference.  I'd recommend that we make orphanTx an object rather than a reference...